floodlight/core/events.py

```diff
@@ -61,3 +74,155 @@ def custom(self):
+        floodlight.core.definitions) for all protected columns in the inner events
+        DataFrame
+
+        Returns
+        -------
+        check: bool
+            True if the checks for all protected columns pass and False otherwise
+        """
+        check = True
+        for col in self.protected:
+            check *= self.definition_check(col, protected_columns)
+
+        return check
+
+    def definition_check(self, col: str, definitions: Dict[str, Dict]) -> bool:
+        """Perform the checks for value range for a single column of the inner event
+        DataFrame using the specifications from floodlight.core.definitions.
+
+        Parameters
+        ----------
+        col: str
+            Column of the inner event DataFrame to be checked
+        definitions: Dict
+            Dictionary for each column name to be checked. Each entry is a Dictionary
+            containing definitions for allowed value ranges of the respective
+            column. Information about value_range is given as a List of the form
+            [min_value, max_value, extra_value] where the extra_value indicates an
+            additionally allowed value that does not fit within the defined value range.
+
+        Returns
+        -------
+        check: bool
+            True if the checks for value range pass and False otherwise
+        """
+        check = True
+
+        # check if value range is defined
+        if definitions[col]["value_range"] is None:
+            return check
+
+        # skip extra allowed values
+        check_col = self.events[col]
+        if len(definitions[col]["value_range"]) == 3:
+            if definitions[col]["value_range"][2] is None:
+                check_col = check_col.dropna()
+            else:
+                check_col = check_col[check_col != definitions[col]["value_range"][2]]
+
+        # check value range for remaining values
+        if not (definitions[col]["value_range"][0] <= check_col).all():
+            check = False
+            warnings.warn(
+                f"Events column {col} has at least one value smaller than "
+                f"{definitions[col]['value_range'][0]}!"
+            )
+        if not (check_col <= definitions[col]["value_range"][1]).all():
+            check = False
+            warnings.warn(
+                f"Events column {col} has at least one value larger than "
+                f"{definitions[col]['value_range'][1]}!"
+            )
+
+        return check
+
+    def add_frameclock(self, framerate: int):
+        """Add the column "frameclock", computed as the rounded multiplication of
+        gameclock and framerate, to the inner events DataFrame.
+
+        Parameters
+        ----------
+        framerate: int
+            Temporal resolution of data in frames per second/Hertz.
+        """
+        frameclock = np.full((len(self.events)), -1, dtype=int)
+        frameclock[:] = np.round(self.events["gameclock"].values * framerate)
```

Comment:
Assume you have framerate of 1, and an event happens after 2.6 seconds. In this case,  np.round(gameclock * framerate) = np.round(2.6) = 3.

I'm inclined to say that we should always use the lower integer, i.e. 2 in this case. If I had to round the time of an event (say 3:56) to a minute, I'd still say the event happend in minute 3. 

If you go for the latter, int() or np.floor() will do the job.

What's your opinion?

Reply:
Plus if you use this in conjunction with a frame-based object, i.e. to filter the positions for a given event, rounding up will give you a snapshot of the data *after* the event happened

Reply:
Although it is awful to round something like 2.99 seconds to frame 2, I agree that we rather want to have the frame being too early than too late. I will change it to int().